spec: update spec to support downstream snapshot group and replication partners

raunakkumar opened the pull request:
Signed-off-by: Raunak <raunak.kumar@hpe.com>

raunakkumar committed:
spec: update csp spec for volumes to return free_bytes and used_bytes

Signed-off-by: Raunak <raunak.kumar@hpe.com>

datamattsson commented on spec.md:
```diff
@@ -154,7 +158,9 @@ GET http://localhost:8080/csp/containers/v1/volumes/067b5b0c6a3d0ece060000000000
-    }
+    },
+    "used_bytes" : 12570624,
+    "free_bytes" : 1061171200,
```
Ending comma?

datamattsson reviewed (approved):
Small nit then LGTM.

datamattsson commented on spec.md:
```diff
@@ -129,7 +131,9 @@ GET http://localhost:8080/csp/containers/v1/volumes
-        }
+        },
+        "used_bytes" : 12570624,
+        "free_bytes" : 1061171200,
```
Ending comma?

raunakkumar replied:
updated

raunakkumar committed:
csp: update spec to support for downstream snapshot group and replication partner

Signed-off-by: Raunak <raunak.kumar@hpe.com>

rgcostea reviewed (commented):
Let's have an internal design review on this proposal

raunakkumar committed:
spec: support replication partner via volumegroup config
- replication partnership was only supported via protectionTemplate,
make it available to be independently set

Signed-off-by: Raunak <raunak.kumar@hpe.com>

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.yaml:
```diff
@@ -790,6 +790,16 @@ definitions:
+        - N/A
+        - 'No'
+        - 'Yes'
+      volumeCollectionList:
```
Should we wait to add this until a use case for it is identified?

raunakkumar replied:
I had it marked as hidden, not sure why it still spits out.  I will remove it for time being as you suggested
```    
@ApiModelProperty(hidden = true)
```

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -245,10 +245,10 @@
       "get" : {
-        "summary" : "Get a snapshot group with the given ID",
-        "operationId" : "getSnapshotGroup",
+        "summary" : "Get all snapshot groups used for container workloads",
```
Should it be "Get the snapshot groups for a particular volume group"?

raunakkumar committed:
update spec to address downstream snapshotgroup deletion

Signed-off-by: Raunak <raunak.kumar@hpe.com>

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.json:
```diff
@@ -375,16 +375,30 @@
+            "schema": {
+              "type": "array",
+              "items": {
+                "$ref": "#/definitions/ContainerVolumeSnapshot"
```
The only thing we need from the ContainerVolumeSnapshot is its ID, right?  Maybe it is enough to just take a list of IDs as input.

How large are the IDs, and how many might there be?  I'm curious how viable it would be to include a comma-separated list in a URI parameter instead of putting them in a request body.

raunakkumar replied:
Snapshot IDs would suffice. I wasn't sure if doing the validation at the client with request body was helpful and went with the request body based on our discussion yesterday. Let me update with just the list of IDs as query params

rkumpf replied:
A URI is limited to 2048 characters.  If the list of snapshot IDs could approach this length, we'll need to put them in the message body.

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -419,13 +419,23 @@
+          "name" : "replication_partner",
           "in" : "query",
-          "description" : "Indicates if this is the downstream snapshot group to be deleted",
+          "description" : "Replication partner name to delete downstream snapshot group",
```
I suggest:  "The name of a downstream replication partner from which to delete the snapshot group, with no effect upstream"

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -459,52 +469,42 @@
           },
           "404" : {
-            "description" : "Not Found if volume ID or snapshot name cannot be found"
+            "description" : "Not Found if snapshot ID cannot be found"
```
I suggest "Snapshot ID not found", or simply "Not Found".

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -517,45 +517,35 @@
+        }, {
+          "name" : "id",
+          "in" : "path",
+          "description" : "The id of the snapshot to delete",
```
Capitalize "ID"

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -517,45 +517,35 @@
             "description" : "Unauthorized"
           },
+          "404" : {
+            "description" : "Not found if snapshot ID is not found"
```
Same comment as at line 498.

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -517,45 +517,35 @@
       "get" : {
-        "summary" : "Get a snapshot with the given ID",
-        "operationId" : "getSnapshot",
+        "summary" : "Get all snapshots for the provided volume",
```
I suggest: "Get snapshots for a specified volume"

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -569,42 +559,52 @@
+        }, {
+          "name" : "name",
+          "in" : "query",
+          "description" : "Snapshot name",
```
I suggest: "The name of a particular snapshot to get"

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -569,42 +559,52 @@
           },
           "404" : {
-            "description" : "Not Found if snapshot ID cannot be found"
+            "description" : "Not Found if volume ID or snapshot name cannot be found"
```
If the snapshot name isn't found, shouldn't we just return an empty list of snapshots?

I realize this only shows up in the diff because of an ordering change, so you can ignore this comment if you like.  The reordering sure makes the PR annoying to review, though.

raunakkumar replied:
Ya, its annoying since we are generating these files from spec.

rgcostea replied:
At this point, we probably don't need to generate from source any longer.  Would it be a huge hassle on the nimble side to start generating stubs from the yaml from now on?

raunakkumar replied:
I will look into it.

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.json:
```diff
@@ -375,21 +375,35 @@
           },
           "404": {
-            "description": "Not found if snapshot group ID is not found"
+            "description": "No Snapshot group ID found"
```
"Snapshot group ID not found"

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.yaml:
```diff
@@ -246,24 +246,33 @@ paths:
           description: Unauthorized
         404:
-          description: Not found if snapshot group ID is not found
+          description: No Snapshot group ID found
```
"Snapshot group ID not found"

rkumpf commented on nimble-storage/nimble-csp-api.json:
```diff
@@ -310,45 +310,52 @@
         } ],
         "requestBody" : {
-          "description" : "A new snapshot group that is to be created",
+          "description" : "List of snapshot IDs of the downstream snapshotGroup",
```
"List of IDs of the snapshots in the downstream snapshot group"

rkumpf reviewed (commented):
Looks good

raunakkumar committed:
csp: remove replicateTo in favor of replicate

Signed-off-by: Raunak <raunak.kumar@hpe.com>

rgcostea commented on spec.md:
```diff
@@ -958,3 +1017,8 @@ DELETE http://localhost:8080/csp/containers/v1/snapshot_groups/052265c9672660666
 | | config | map[string]interface{} | | X | X |
+| ReplicationPartner | | | | | |
+| | id | string | | | X |
+| | partner_name | string | X |  | X |
```
Can this just be name?

raunakkumar replied:
sure. Will update

raunakkumar committed:
spec: update examples

Signed-off-by: Raunak <raunak.kumar@hpe.com>

raunakkumar commented:
@rgcostea Is this PR good to be merged? 
I've looked into swagger-codegen to generate the object/objectsets by using the cli but it doesn't generate the code with all annotations in the code. I will look further into swagger-codegen gradle plugin if we can get further.
```
 java -jar swagger-codegen-cli.jar generate -i nimble-csp-api-swagger-v2.json -l java
```

rgcostea commented on spec.md:
````diff
@@ -873,6 +883,51 @@ DELETE http://localhost:8080/csp/containers/v1/snapshot_groups/052265c9672660666
 204 No Content
 ```
 
+DELETE `/containers/v1/snapshot_groups?delete_downstream=true`
+
+ * Delete the downstream snapshotgroup 
````
This doesn't look right.  It should be /containers/v1/snapshot_groups/{id}

rgcostea commented on spec.md:
````diff
@@ -873,6 +883,51 @@ DELETE http://localhost:8080/csp/containers/v1/snapshot_groups/052265c9672660666
 204 No Content
 ```
 
+DELETE `/containers/v1/snapshot_groups?delete_downstream=true`
````
Does this only delete the downstream snapshot_group?

raunakkumar replied:
aah, missed to update the spec. This is the final decision we went with
```          
@Parameter(
                    description = "Authorization headers",
                    required = true)
            @ApiParam(
                    value = "Authorization headers",
                    hidden = true
            )
            @BeanParam RequestBean request,

            @Parameter(
                    description = "The id of the snapshot group to delete",
                    required = true)
            @PathParam("id")
                    String id,
            @Parameter(
                    description = "The name of a downstream replication partner from which to delete the snapshot group, with no effect upstream")
            @QueryParam("replication_partner")
                    String replicationPartner,
            @RequestBody(
                    description = "List of IDs of the snapshots in the downstream snapshot group",
                    content = @Content(
                            schema = @Schema(implementation = SnapshotGroup.class)))
                    List<String> snapshotIds)
```

rgcostea replied:
Why does the replication_partner need to be specified?  Can't we just look that up from the nimble array?

raunakkumar replied:
We want to support deletion of remote snapshots independently and we need the list of snapshot ids and replication partner to achieve this. 
There is a possibility that the upstream snapshots could be deleted while only the downstream remains. Since the volumegroup (added to spec) and snapshotgroup already know about the replication partner and snapshot ids, we don't have to make additional calls

rgcostea replied:
Hmmm.  Why can't the API do the following:

```
SnapshotCollection snapColl = NimOS.getSnapshotCollection(id);
String partnerName = snapColl.getReplicateTo();
String partnerId = NimOS.getReplicationpartnerIdFromName(partnerName);
for (String snapshotId : sg.getSnapshotList()) {
    NimOS.deleteSnapshot(snapshotID, partnerID);
}
```

raunakkumar replied:
```
snapColl = NimOS.getSnapshotCollection(id);
```
this would work if the upstream snapcoll still exists. We identified a scenario where its possible for upstream and downstream to be independent.

rgcostea replied:
Please elaborate on that scenario.  Upstream and downstream snapcolls are always connected AFAIK.  Otherwise, what's the purpose of supplying the id of the snapshot_group?  Why not just do "DELETE /containers/v1/snapshots/{id}?delete_downstream=true"?

raunakkumar replied:
The lifecycle of the upstream/downstream snapcolls are independent after the replication is complete. 
We can still have downstream snapshots even though the upstream has been deleted (we would want to support this use case from our design discussions).
Now, to be able to delete downstream snapcolls from an upstream snapshots we can still achieve it if we feed it with enough information for replication_partner to invoke the delete on the downstream.
The pre-requisite being that the replication relationship still exists.
I have covered this in the Nimble CSP SnapshotGroupTest scenario as followis
```
        1. delete the local snapshotGroup
        2. the upstream snapshotgroup should have been deleted
        3. now delete the remote snapshotGroup 
        4. the delete should go through with the partner_name and list of snapshots.
```

rgcostea commented:
Have we confirmed with the Primera team that this spec proposal is something they can implement?

Can they delete the downstream from the upstream?  If not, then I don't think any changes to the spec are needed.  We would just execute DELETE /containers/v1/snapshot_groups/{id} against the downstream array.

raunakkumar commented:
> Have we confirmed with the Primera team that this spec proposal is something they can implement?
> 
> Can they delete the downstream from the upstream? If not, then I don't think any changes to the spec are needed. We would just execute DELETE /containers/v1/snapshot_groups/{id} against the downstream array.

They have yet to look into it but we decided that if we could support remote snapshots deletion from upstream we should allow for it although it can be achieved by the mechanism you mentioned. This gives us a flexibility.

raunakkumar committed:
spec: remove deletion of downstream snapshotgroups from upstream

Signed-off-by: Raunak <raunak.kumar@hpe.com>

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.json:
```diff
@@ -1143,6 +1193,32 @@
+        "repl_direction": {
+          "type": "string",
+          "enum": [
+            "upstream",
```
Minor comment...  The values "upstream" and "downstream" don't seem right for a "replication direction".  Replication always goes from upstream to downstream.  Consider calling it "repl_role" or "repl_position" instead of "repl_direction".

Also change "bi_directional" to "bidirectional".

raunakkumar replied:
These are popping in from some of the existing NimOS definitions.

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.json:
```diff
@@ -1271,11 +1372,42 @@
         "config": {
           "$ref": "#/definitions/VolumeConfig"
         },
+        "snaphotUsedBytes": {
```
Did these sneak back in?

rkumpf commented on nimble-storage/nimble-csp-api-swagger-v2.yaml:
```diff
@@ -116,6 +117,35 @@ paths:
+        401:
+          description: Unauthorized
+        404:
+          description: Replication partner name cannot be found
```
Change "cannot be" to "not".

raunakkumar committed:
spec: update replDirection to replRole

Signed-off-by: Raunak <raunak.kumar@hpe.com>

rkumpf commented:
It looks good except for the VolumeConfig properties which might have been unintentionally (re)added.

raunakkumar commented:
@rgcostea Is the PR good to be merged now?

rgcostea commented:
> @rgcostea Is the PR good to be merged now?

I think this should be versioned.  The 1.0.0 version of the spec does not required volume_groups and snapshot_groups to be implemented.  We should version that as 1.1.0 and then commit your latest updates to master.

raunakkumar commented:
Thanks

raunakkumar commented:
> > @rgcostea Is the PR good to be merged now?
> 
> I think this should be versioned. The 1.0.0 version of the spec does not required volume_groups and snapshot_groups to be implemented. We should version that as 1.1.0 and then commit your latest updates to master.

Oops. I will raise another PR to reflect the version update
